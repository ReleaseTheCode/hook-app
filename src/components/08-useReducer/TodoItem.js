import React from 'react';
import PropTypes from 'prop-types';

export const TodoItem = ({ index, todo, handleToggle, handleDelete}) => {
    return (
        <li 
            key={ todo.id }
            className='list-group-item'
        >
            <p 
                className={`${todo.done && 'complete' }`}
                onClick={ () => handleToggle( todo.id ) }
            > 
                { index} { todo.desc }
            </p>
            <button className='btn btn-danger' onClick={()=>handleDelete(todo.id)} >Borrar</button>
        </li>
    )
}

TodoItem.propTypes = {
    index: PropTypes.number,
    todo: PropTypes.object,
    handleToggle: PropTypes.func,
    handleDelete: PropTypes.func
}