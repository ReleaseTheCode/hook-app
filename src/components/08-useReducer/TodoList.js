import React from 'react'
import PropTypes from 'prop-types';
import { TodoItem } from './TodoItem'

export const TodoList = ( { todos, handleToggle, handleDelete } ) => {
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map( ( todo, index ) => (
                    <TodoItem
                        key={ todo.id }
                        index={ index+1 }
                        todo={todo}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                ) )
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array,
    handleToggle: PropTypes.func,
    handleDelete: PropTypes.func
}