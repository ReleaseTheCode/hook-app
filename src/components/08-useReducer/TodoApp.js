import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import './styles.css'
import { useForm } from '../../hooks/useForm'
import { TodoList } from './TodoList';

const init = () => {
    return JSON.parse( localStorage.getItem('todos') ) || [];
    // [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false
    // }];
}

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer( todoReducer, [], init );
    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });
    
    console.log(todos);
    
    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify( todos ) );
    }, [todos]);

    const handleDelete = ( todoId ) => {
        const action = {
            type: 'delete',
            payload: todoId
        };
        dispatch( action );
    }

    const handleToggle = ( todoId ) => {
        dispatch ({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if( description.trim().length <= 1 )
            return;

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        const action = {
            type: 'add',
            payload: newTodo
        };

        dispatch( action );
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({ todos.length })</h1>
            <hr/>
            <div className='row'>
                <div className='col-7'>
                    <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete}/>
                </div>
                <div className='col-5'>
                    <h4>Agregar TODO</h4>
                    <hr/>
                    <form className='d-grid' onSubmit={ handleSubmit }>
                        <input
                            type='text'
                            name='description'
                            className='form-control'
                            placeholder='Aprender...'
                            autoComplete='off'
                            value={ description }
                            onChange={ handleInputChange }
                        />
                        <button 
                            type='submit'
                            className='btn btn-outline-primary mt-1'>Agregar</button>
                    </form>
                </div>
            </div>
            

        </div>
    )
};
