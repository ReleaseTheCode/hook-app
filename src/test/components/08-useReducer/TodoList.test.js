import React from 'react'
import { shallow } from "enzyme"
import { TodoList } from "../../../components/08-useReducer/TodoList"
import { demoTodos } from '../../fixtures/demoTodos'

describe('Test on <TodoList/>', () => {
const handleToggle = jest.fn();
const handleDelete = jest.fn();
    const wrapper = shallow(
        <TodoList
            todos={ demoTodos }
            handleToggle={handleToggle}
            handleDelete={handleDelete}
        />
     );
    test('should render', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should have TodoItem = demoTodos.length ', () => {
        expect( wrapper.find('TodoItem').length ).toBe( demoTodos.length );
        expect( wrapper.find('TodoItem').at(0).prop('handleDelete') ).toEqual( expect.any(Function) );
        expect( wrapper.find('TodoItem').at(0).prop('handleToggle') ).toEqual( expect.any(Function) );
    });
    
    
})
