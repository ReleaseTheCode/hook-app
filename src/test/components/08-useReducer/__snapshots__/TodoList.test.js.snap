// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoList/> should render 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="1"
    todo={
      Object {
        "desc": "Comprar pan",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={2}
    key="2"
    todo={
      Object {
        "desc": "Comprar leche",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
